### PR TITLE
Handle purl fragments in package search #1032

### PR DESCRIPTION
**Pull request by TG1999**
Signed-off-by: Tushar Goel <tushar.goel.dav@gmail.com>

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -2,6 +2,11 @@ Release notes
 =============
 
 
+Version v31.1.0
+----------------
+
+- We are now handling purl fragments in package search.
```
Comment (pombredanne) [inline CHANGELOG.rst:8]: Please elaborate on what this means for a user

File: vulnerabilities/models.py [hunk 2]
```diff
@@ -412,61 +412,22 @@ def search(self, query=None):
...
-                qs = qs.filter(version__iexact=version)
-
-        return qs
+
+        try:
+            # if it's a valid purl, use it as is
+            purl = PackageURL.from_string(query)
+            return self.for_purl(purl, with_qualifiers_and_subpath=False)
+        except ValueError:
+            return qs.filter(package_url__icontains=query)
```
Comment (pombredanne) [inline vulnerabilities/models.py:422]: Are you sure this provides the same kind of search results as before?

File: vulnerabilities/models.py [hunk 3]
```diff
@@ -412,61 +412,22 @@ def search(self, query=None):
...
 
     def for_purl(self, purl, with_qualifiers_and_subpath=True):
         """
         Return a queryset matching the ``purl`` Package URL.
         """
         if not isinstance(purl, PackageURL):
             purl = PackageURL.from_string(purl)
-        purl = purl.to_dict()
+        purl = purl_to_dict(purl)
```
Comment (pombredanne) [inline vulnerabilities/models.py:430]: What is purl_to_dict() doing?

File: vulnerabilities/tests/test_view.py [hunk 4]
```diff
@@ -57,6 +57,12 @@ def test_package_view(self):
         self.assertEqual(len(pkgs), 1)
         self.assertEqual(pkgs[0].purl, "pkg:nginx/nginx@1.0.15")
 
+    def test_package_view_with_purl_fragment(self):
```
Comment (pombredanne) [inline vulnerabilities/tests/test_view.py:60]: We need many more tests IMHO to assert the search results quality before and ensure this does not regress.

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks. Some nits for your review.

File: vulnerabilities/tests/test_view.py (hunk 4 above)
Comment (TG1999) [reply vulnerabilities/tests/test_view.py:60]: @pombredanne we already had tests for the previous implementation and none of them fail, please tell me any particular corner case that should we check for ?

File: vulnerabilities/models.py (hunk 3 above)
Comment (TG1999) [reply vulnerabilities/models.py:430]: This returns a dict of purl components for use in the queryset.

**Review by pombredanne — APPROVED** [review]
LGTM! Thanks.
